lib/winapi/timezoneapi.c

```diff
@@ -0,0 +1,196 @@
+
+typedef struct {
+    UCHAR Month;     // 1-12
+    UCHAR Day;       // 1 = 1st occurrence of DayOfWeek up to 5 (5th or last)
+    UCHAR DayOfWeek; // 0 = Sunday to 6 = Saturday
+    UCHAR Hour;
+} XBOX_TZ_STRUCT;
+
+// Determine the day of the week given a year, month and day
+// https://en.wikipedia.org/wiki/Determination_of_the_day_of_the_week#Methods_in_computer_code
+static UCHAR GetDayofWeek(SHORT y, UCHAR m, UCHAR d) {
+    return (d += m < 3 ? y-- : y - 2, 23 * m / 9 + d + 4 + y / 4 - y / 100 + y / 400) % 7;
+}
+
+static UCHAR GetDaysinMonth(SHORT year, UCHAR month) {
+    static const UCHAR daysPerMonth[] = {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
+    UCHAR days;
+
+    assert (month <= 12);
+
+    if (month == 2) {
+        if ((year % 4 == 0 && year % 100 != 0) || (year % 400 == 0)) {
+            days = 29;
+        } else {
+            days = 28;
+        }
+    } else {
+        days = daysPerMonth[month - 1];
+    }
+
+    return days;
+}
+
+// Given a year, month, day of the week, and the nth occurrence this returns the day in the month.
+// i.e The 2nd wednesday in November 2023 = 8th November. This function would return 8.
+static UCHAR GetActualDay(SHORT year, XBOX_TZ_STRUCT *tzInfo)
+{
+    UCHAR dayOfWeek = GetDayofWeek(year, tzInfo->Month, 1);
+    UCHAR daysinMonth = GetDaysinMonth(year, tzInfo->Month);
+    UCHAR dayTargetCnt = tzInfo->Day - 1;
+    UCHAR dayActual = 1;
+
+    // Find first occurrence of the weekday
+    dayActual += (tzInfo->DayOfWeek < dayOfWeek) ? (tzInfo->DayOfWeek + 7 - dayOfWeek) : (tzInfo->DayOfWeek - dayOfWeek);
+
+    // Add remaining weeks
+    dayActual += dayTargetCnt * 7;
+
+    // If overrun, go back to the last occurrence in the month
+    while (dayActual > daysinMonth)
+    {
+        dayActual -= 7;
+    }
+    return dayActual;
+}
+
+DWORD GetTimeZoneInformation(LPTIME_ZONE_INFORMATION lpTimeZoneInformation)
+{
+    assert(lpTimeZoneInformation != NULL);
+
+    LONG timeZoneBias, dayLightBias, standardBias, dayLightEnFlag;
+    LONG timeNow, startDST, endDST;
+    XBOX_TZ_STRUCT dayLightDate, standardDate;
+    CHAR dayLightName[4], standardName[4];
+    NTSTATUS queryStatus;
+    TIME_FIELDS dateNow;
+    LARGE_INTEGER kTime;
+    ULONG type;
+
+    RtlZeroMemory(lpTimeZoneInformation, sizeof(TIME_ZONE_INFORMATION));
+
+    queryStatus = ExQueryNonVolatileSetting(XC_TIMEZONE_BIAS, &type, &timeZoneBias, sizeof(timeZoneBias), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+
+    lpTimeZoneInformation->Bias = timeZoneBias;
+
+    // Check if DST adjustment is enabled
+    queryStatus = ExQueryNonVolatileSetting(XC_MISC, &type, &dayLightEnFlag, sizeof(dayLightEnFlag), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+
+    // DST adjustment is disabled.
```

Comment:
Some of the comments end with a period symbol, others don't. Copy/paste bug?

Reply:
I just removed all full stops so its consistent. Looking at some other files we didnt seem to use them often.